File: sdks/python/speechswitch/protobuf.py
"""Scalar protobuf primitives; generated clients own field numbers and shapes."""

import builtins
import math
import struct


class ProtoWriter:
    def __init__(self) -> None:
        self._data = bytearray()

    def uint32(self, value: object) -> "ProtoWriter":
        if type(value) is not int or not 0 <= value <= 0xFFFFFFFF:
            raise TypeError("Invalid protobuf uint32")
        return self._varint(value)

    def int32(self, value: object) -> "ProtoWriter":
        if type(value) is not int or not -2147483648 <= value <= 2147483647:
            raise TypeError("Invalid protobuf int32")
        return self._varint(value & 0xFFFFFFFFFFFFFFFF)

    def _varint(self, value: int) -> "ProtoWriter":
        while value > 127:
            self._data.append((value & 127) | 128)
            value >>= 7
        self._data.append(value)
        return self

    def bool(self, value: object) -> "ProtoWriter":
        if type(value) is not bool:
            raise TypeError("Invalid protobuf bool")
        return self.uint32(1 if value else 0)

    def double(self, value: object) -> "ProtoWriter":
        if not isinstance(value, (int, float)) or isinstance(value, bool):
            raise TypeError("Invalid protobuf double")
        try:
            valid = math.isfinite(value)
        except OverflowError:
            valid = False
        if not valid:
            raise TypeError("Invalid protobuf double")
        self._data.extend(struct.pack("<d", value))
        return self

    def bytes(self, value: object) -> "ProtoWriter":
        if not isinstance(value, bytes):
            raise TypeError("Invalid protobuf bytes")
        self.uint32(len(value))
        self._data.extend(value)
        return self

    def string(self, value: object) -> "ProtoWriter":
        if not isinstance(value, str):
            raise TypeError("Invalid protobuf string")
        try:
            encoded = value.encode("utf-8")
        except UnicodeEncodeError:
            # Match TextEncoder: combine surrogate pairs and replace lone units.
            encoded = value.encode("utf-16-le", "surrogatepass").decode("utf-16-le", "replace").encode("utf-8")
        return self.bytes(encoded)

    def finish(self) -> builtins.bytes:
        return bytes(self._data)


class ProtoReader:
    def __init__(self, data: object) -> None:
        if not isinstance(data, builtins.bytes):
            raise TypeError("Invalid protobuf bytes")
        self._data = data
        self._offset = 0

    @property
    def done(self) -> bool:
        return self._offset == len(self._data)

    def _take(self, length: int) -> builtins.bytes:
        if length > len(self._data) - self._offset:
            raise TypeError("Truncated protobuf message")
        data = self._data[self._offset:self._offset + length]
        self._offset += length
        return data

    def _varint(self) -> int:
        result = 0
        for index in range(10):
            byte = self._take(1)[0]
            if index == 9 and byte > 1:
                raise TypeError("Invalid protobuf varint")
            result |= (byte & 127) << (index * 7)
            if byte & 128 == 0:
                return result
        raise TypeError("Invalid protobuf varint")

    def uint32(self) -> int:
        value = self._varint()
        if value > 0xFFFFFFFF:
            raise TypeError("Invalid protobuf uint32")
        return value

    def int32(self) -> int:
        value = self._varint() & 0xFFFFFFFF
        return value if value < 0x80000000 else value - 0x100000000

    def bool(self) -> builtins.bool:
        return self._varint() != 0

    def double(self) -> float:
        return float(struct.unpack("<d", self._take(8))[0])

    def bytes(self) -> builtins.bytes:
        return self._take(self.uint32())

    def string(self) -> str:
        return self.bytes().decode("utf-8", errors="strict")

    def skip(self, wire: int) -> None:
        if wire == 0:
            self._varint()
        elif wire == 1:
            self._take(8)
        elif wire == 2:
            self.bytes()
        elif wire == 5:
            self._take(4)
        else:
            raise TypeError(f"Unsupported protobuf wire type: {wire}")

File: sdk/runtime/protobuf.ts
/** Scalar protobuf primitives; generated clients supply field numbers and types. */
export class ProtoWriter {
  private readonly chunks: Uint8Array[] = [];
  uint32(value: number): this {
    if (!Number.isInteger(value) || value < 0 || value > 0xffffffff) throw new TypeError("Invalid protobuf uint32");
    const bytes: number[] = [];
    do { const byte = value % 128; value = Math.floor(value / 128); bytes.push(byte | (value ? 128 : 0)); } while (value);
    this.chunks.push(Uint8Array.from(bytes)); return this;
  }
  int32(value: number): this {
    if (!Number.isInteger(value) || value < -2147483648 || value > 2147483647) throw new TypeError("Invalid protobuf int32");
    if (value >= 0) return this.uint32(value);
    let wide = BigInt.asUintN(64, BigInt(value)); const bytes: number[] = [];
    do { const byte = Number(wide & 127n); wide >>= 7n; bytes.push(byte | (wide ? 128 : 0)); } while (wide);
    this.chunks.push(Uint8Array.from(bytes)); return this;
  }
  bool(value: boolean): this {
    if (typeof value !== "boolean") throw new TypeError("Invalid protobuf bool");
    return this.uint32(value ? 1 : 0);
  }
  double(value: number): this {
    if (!Number.isFinite(value)) throw new TypeError("Invalid protobuf double");
    const bytes = new Uint8Array(8); new DataView(bytes.buffer).setFloat64(0, value, true);
    this.chunks.push(bytes); return this;
  }
  bytes(value: Uint8Array): this {
    if (!(value instanceof Uint8Array)) throw new TypeError("Invalid protobuf bytes");
    this.uint32(value.byteLength); this.chunks.push(value); return this;
  }
  string(value: string): this {
    if (typeof value !== "string") throw new TypeError("Invalid protobuf string");
    return this.bytes(new TextEncoder().encode(value));
  }
  finish(): Uint8Array<ArrayBuffer> {
    const bytes = new Uint8Array(this.chunks.reduce((size, chunk) => size + chunk.byteLength, 0));
    let offset = 0; for (const chunk of this.chunks) { bytes.set(chunk, offset); offset += chunk.byteLength; }
    return bytes;
  }
}

export class ProtoReader {
  private offset = 0;
  private readonly data: Uint8Array;
  constructor(data: Uint8Array) { this.data = data; }
  get done(): boolean { return this.offset === this.data.byteLength; }
  private take(length: number): Uint8Array {
    if (length > this.data.byteLength - this.offset) throw new TypeError("Truncated protobuf message");
    const value = this.data.subarray(this.offset, this.offset + length); this.offset += length; return value;
  }
  private varint(): bigint {
    let result = 0n;
    for (let index = 0; index < 10; index++) {
      const byte = this.take(1)[0]!;
      if (index === 9 && byte > 1) throw new TypeError("Invalid protobuf varint");
      result |= BigInt(byte & 127) << BigInt(index * 7);
      if (!(byte & 128)) return result;
    }
    throw new TypeError("Invalid protobuf varint");
  }
  uint32(): number {
    const value = this.varint();
    if (value > 0xffffffffn) throw new TypeError("Invalid protobuf uint32");
    return Number(value);
  }
  int32(): number { return Number(BigInt.asIntN(32, this.varint())); }
  bool(): boolean { return this.varint() !== 0n; }
  double(): number {
    const bytes = this.take(8); return new DataView(bytes.buffer, bytes.byteOffset, 8).getFloat64(0, true);
  }
  bytes(): Uint8Array { return this.take(this.uint32()); }
  string(): string { return new TextDecoder("utf-8", { fatal: true }).decode(this.bytes()); }
  skip(wire: number): void {
    switch (wire) {
      case 0: this.varint(); return;
      case 1: this.take(8); return;
      case 2: this.bytes(); return;
      case 5: this.take(4); return;
      default: throw new TypeError(`Unsupported protobuf wire type: ${wire}`);
    }
  }
}
